pkg/controllermanager/controller/project/resourcequota/add.go

```diff
@@ -0,0 +1,103 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package resourcequota
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "project-resourcequota"
+
+var (
+	createOnlyPredicate = predicate.Funcs{
+		CreateFunc:  func(_ event.CreateEvent) bool { return true },
+		DeleteFunc:  func(_ event.DeleteEvent) bool { return false },
+		UpdateFunc:  func(_ event.UpdateEvent) bool { return false },
+		GenericFunc: func(_ event.GenericEvent) bool { return false },
+	}
+)
+
+// AddToManager adds a controller with the given Options to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.ConcurrentSyncs, 0),
+			ReconciliationTimeout:   controllerutils.DefaultReconciliationTimeout,
+		}).
+		For(&corev1.ResourceQuota{}, builder.WithPredicates(r.ObjectInProjectNamespace())).
+		Watches(
+			&v1beta1.Shoot{},
+			handler.EnqueueRequestsFromMapFunc(r.MapShootToResourceQuotasInProject(mgr.GetLogger().WithValues("controller", ControllerName))),
+			builder.WithPredicates(createOnlyPredicate),
+		).
+		Complete(r)
+}
+
+// ObjectInProjectNamespace returns a predicate that filters objects that are in Project namespaces.
+func (r *Reconciler) ObjectInProjectNamespace() predicate.Predicate {
+	objectInNamespaceFunc := func(namespace string) bool {
+		ctx := context.Background()
+		project, err := gardenerutils.ProjectForNamespaceFromReader(ctx, r.Client, namespace)
+		return err == nil && project != nil
+	}
+
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return objectInNamespaceFunc(e.Object.GetNamespace())
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return objectInNamespaceFunc(e.Object.GetNamespace())
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return objectInNamespaceFunc(e.ObjectNew.GetNamespace())
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return objectInNamespaceFunc(e.Object.GetNamespace())
+		},
+	}
```

Comment:
```suggestion
	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
		return objectInNamespaceFunc(obj.GetNamespace())
	})
```

You could also drop the `objectInNamespaceFunc` function now